tests/k8s: fix kbs installation on Azure AKS

wainersm opened the pull request:
The Azure AKS addon-http-application-routing add-on is deprecated and cannot be enabled on new clusters which has caused some CI jobs to fail.

Migrated our code to use approuting instead. Unlike addon-http-application-routing, this add-on doesn't configure a managed cluster DNS zone, but the created ingress has a public IP. To avoid having to deal with DNS setup, we will be using that address from now on. Thus, some functions no longer used are deleted.

Fixes #11156
Signed-off-by: Wainer dos Santos Moschetta <wainersm@redhat.com>

---

This is how I tested it locally:

```
$ export KBS_INGRESS=aks
$ ./tests/integration/kubernetes/gha-run.sh create-cluster
<output omitted>
$ ./tests/integration/kubernetes/gha-run.sh get-cluster-credentials
<output omitted>
$ ./tests/integration/kubernetes/gha-run.sh deploy-coco-kbs
<output omitted>
::endgroup::
/tmp/trustee/kbs/config/kubernetes/overlays /tmp/trustee/kbs/config/kubernetes ~/src/github.com/kata-containers/kata-containers
::group::/tmp/trustee/kbs/config/kubernetes/overlays/ingress.yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: kbs
  namespace: coco-tenant
  annotations:
    kubernetes.io/ingress.class: webapprouting.kubernetes.azure.com
spec:
  rules:
  - host: ""
    http:
      paths:
      - path: /
        pathType: Prefix
        backend:
          service:
            name: kbs
            port:
              number: 8080
::endgroup::
/tmp/trustee/kbs/config/kubernetes ~/src/github.com/kata-containers/kata-containers
::group::Deploy the KBS
namespace/coco-tenant created
configmap/kbs-config-d655m9k8gk created
configmap/policy-config-tf9h6d9bb2 created
secret/kbs-auth-public-key-t2f5fd6t89 created
secret/keys-gmf2g75547 created
service/kbs created
deployment.apps/kbs created
Warning: annotation "kubernetes.io/ingress.class" is deprecated, please use 'spec.ingressClassName' instead
ingress.networking.k8s.io/kbs created
~/src/github.com/kata-containers/kata-containers
::endgroup::
::group::Check the service healthy
pod/kbs-checker-767919 created
pod "kbs-checker-767919" deleted
KBS service respond to requests
::endgroup::
::group::Check the kbs service is exposed
Trying to connect at http://51.8.197.252:80. Timeout=350
KBS service respond to requests at http://51.8.197.252:80
::endgroup::
```

The `Warning: annotation "kubernetes.io/ingress.class" is deprecated, please use 'spec.ingressClassName' instead` warning I've a fix that I will send to trustee. This doesn't break the installation, so we don't need to bump the kbs version used here.

wainersm commented:
Hi @Xynnn007 , let's see if it fixes in the CI.

sprt commented:
JFYI I merged #11157

wainersm commented:
Latest CI execution showed that this fixed the kbs installation, however the job got "red" because tests failed. So the code is ready for review.

In meanwhile I will re-trigger the job, see if the tests failures were caused by infra or not.

wainersm commented:
KBS is installed correctly but tests still failing. I cannot think in a reason why my changes would break the tests, so it seems some instability on CI.

I suggest to merge this PR then keep on eyes on the next nightly jobs.

sprt commented:
I wouldn't be so quick to dismiss this as infra instability - I'm seeing tests failing consistently that I haven't seen before.

It seems like this is the common failure mode now:

https://github.com/kata-containers/kata-containers/actions/runs/14625338669/job/41036764485?pr=11164#step:17:5498
```
#   Warning  FailedScheduling  90s   default-scheduler  0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
```

Also this seems suspicious:

https://github.com/kata-containers/kata-containers/actions/runs/14625338669/job/41036764485?pr=11164#step:17:5498
```
app-routing-system   nginx-6564f49b47-c6xcs                          1/1     Running   0          23m
app-routing-system   nginx-6564f49b47-tdbd7                          0/1     Pending   0          23m
```

* Two nginx pods in this app-routing-system namespace - are we creating these?
* One of them is in the pending state

Is it possible that this PR is introducing new control plane containers that eat into our CPU limits?

wainersm commented:
Hi @sprt !

> I wouldn't be so quick to dismiss this as infra instability - I'm seeing tests failing consistently that I haven't seen before.
>

You are right! Thanks for the heads up!
 
> It seems like this is the common failure mode now:
> 
> https://github.com/kata-containers/kata-containers/actions/runs/14625338669/job/41036764485?pr=11164#step:17:5498
> 
> ```
> #   Warning  FailedScheduling  90s   default-scheduler  0/1 nodes are available: 1 Insufficient cpu. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod.
> ```
> 
> Also this seems suspicious:
> 
> https://github.com/kata-containers/kata-containers/actions/runs/14625338669/job/41036764485?pr=11164#step:17:5498
> 
> ```
> app-routing-system   nginx-6564f49b47-c6xcs                          1/1     Running   0          23m
> app-routing-system   nginx-6564f49b47-tdbd7                          0/1     Pending   0          23m
> ```
> 
>     * Two nginx pods in this app-routing-system namespace - are we creating these?
>

 I followed the instructions in https://learn.microsoft.com/en-us/azure/aks/app-routing , and essentially I'm creating the ingress resource with `ingressClassName` set to  `webapprouting.kubernetes.azure.com` likewise.

It will create a nginx pods as a result, under the hoods. I don't know why the pending one.

>     * One of them is in the pending state
> 
> 
> Is it possible that this PR is introducing new control plane containers that eat into our CPU limits?

It is a sound hypothesis. I tested with the azure subscription granted to me by Red Hat, and I didn't notice that problem. Well, I also didn't bother to run the tests either... let me put this PR on draft and see if I can reproduce locally. In meanwhile, if you have any tips about the problem and solution, please, talk to your little brother here ;)

wainersm commented:
Ok, after further investigation my findings are:
 * the app-rounting-system add-on creates a managed nginx ingress controller
 * it's deployed an auto-escaler configured for the minimum of  2 replicas. That's why in https://github.com/kata-containers/kata-containers/pull/11164#issuecomment-2825375130 we see two nginx pods
    * I can manage to set 1 replica by patching the controller:
    ```
   kubectl patch nginxingresscontroller/default -n app-routing-system --type=json -p='[{"op": "add", "path": "/spec/scaling/minReplicas", "value": 1}]'
   ```
   * More info about the controller can be found [here](https://learn.microsoft.com/en-us/azure/aks/app-routing-nginx-configuration?tabs=azurecli)
 * Each nginx pod request 500m of cpu, which is too much for our cluster (single node). Reduce the replicas to one helps but it's still too much. Here is the problem: I could not find a way to lower the cpu requirements via controller. I can edit/patch the `deployment/nginx` object however the controller re-creates it with the default settings (i.e. new pods still get 500m of cpu)

I'm trying to avoid to deploy our own nginx ingress, but honestly I'm not seeing many alternatives. Ideas?

sprt commented:
(Heads up: changing *any* label will trigger a rerun of the CI, so we're seeing different results right now.)

> * Each nginx pod request 500m of cpu, which is too much for our cluster (single node). Reduce the replicas to one helps but it's still too much. Here is the problem: I could not find a way to lower the cpu requirements via controller. I can edit/patch the `deployment/nginx` object however the controller re-creates it with the default settings (i.e. new pods still get 500m of cpu)
> 
> I'm trying to avoid to deploy our own nginx ingress, but honestly I'm not seeing many alternatives. Ideas?

If I'm understanding correctly, the added nginx pod maxes out the node CPU capacity. How about moving the affected tests to the `normal` VM SKU, instead of `small`?

wainersm commented:
> (Heads up: changing _any_ label will trigger a rerun of the CI, so we're seeing different results right now.)
> 
> > * Each nginx pod request 500m of cpu, which is too much for our cluster (single node). Reduce the replicas to one helps but it's still too much. Here is the problem: I could not find a way to lower the cpu requirements via controller. I can edit/patch the `deployment/nginx` object however the controller re-creates it with the default settings (i.e. new pods still get 500m of cpu)
> > 
> > I'm trying to avoid to deploy our own nginx ingress, but honestly I'm not seeing many alternatives. Ideas?
> 
> If I'm understanding correctly, the added nginx pod maxes out the node CPU capacity. How about moving the affected tests to the `normal` VM SKU, instead of `small`?

You got it correctly. Either we (somehow) lower the cpu requested of each nginx pod or increase the worker node size. I don't know how to lower the cpu requests of the azure managed nginx ingress, so we can try to increase the VM sku as you suggested @sprt. How? Is it configured on the subscription?

sprt commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -547,6 +551,18 @@ _handle_ingress_nodeport() {
+	local ingress="${1:-}"
+
+	if [[ "${ingress}" = "aks" ]]; then
+		echo "Patch the ingress controller to have only one replica of nginx"
```
I'd add a one-line comment to quickly explain why this is needed 🙂

wainersm commented:
I've just added a few commits.... @sprt 's comment sparked some ideas. Here, I'm now exporting `K8S_TEST_HOST_TYPE=all` and tweaked the script to create a cluster with larger ("normal") VMs. One side effect of this, is on coco non-tee job it will begin to run a few tests from the "normal" subset, which increases coverage. Also I added the code to lower the amount of nginx replicas to one, although with a bigger cluster it would be able to launch the default two replicas.

As I changed the workflow, I pushed the changes to a topic branch to run tests...as soon I get the results I will post here.

wainersm commented:
In https://github.com/kata-containers/kata-containers/actions/runs/14713221646/job/41292948604 kBS got installed and all tests except for pid-ns passed. I'm going to spin the job again and see if same test fail.

sprt commented:
@wainersm Seeing a bunch of failures here:

https://github.com/kata-containers/kata-containers/actions/runs/14713397922/job/41292694318?pr=11164#step:17:11250

```
[run_kubernetes_tests.sh:147] ERROR: Tests FAILED from suites: k8s-cpu-ns.bats k8s-credentials-secrets.bats k8s-limit-range.bats k8s-pod-quota.bats k8s-qos-pods.bats k8s-nginx-connectivity.bats
```

wainersm commented:
> @wainersm Seeing a bunch of failures here:
> 
> https://github.com/kata-containers/kata-containers/actions/runs/14713397922/job/41292694318?pr=11164#step:17:11250
> 
> ```
> [run_kubernetes_tests.sh:147] ERROR: Tests FAILED from suites: k8s-cpu-ns.bats k8s-credentials-secrets.bats k8s-limit-range.bats k8s-pod-quota.bats k8s-qos-pods.bats k8s-nginx-connectivity.bats
> ```

Please disregard the jobs triggered on this PR. The change on the workflow matter for those tests stop failing, so I've tested using the `topic/wainersm-fix_kbs_on_aks` topic branch.  The latest and greatest job is https://github.com/kata-containers/kata-containers/actions/runs/14713221646/job/41305482015

wainersm commented:
ok, the last job succeeded: https://github.com/kata-containers/kata-containers/actions/runs/14713221646/job/41305482015

It's ready to be merged if no other comments. In meanwhile I will re-trigger the job one more time just for the case...

stevenhorsman commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -419,13 +419,20 @@ function kbs_k8s_deploy() {
 }
 
-# Return the kbs service host name in case ingress is configured
+# Return the kbs service public IP in case ingress is configured
```
nit: Should we make this "echos the..." as (unfortunately) we can't return a string value here?

stevenhorsman commented on tests/integration/kubernetes/confidential_kbs.sh:
```diff
@@ -419,13 +419,20 @@ function kbs_k8s_deploy() {
+		while true; do
+			host=$(kubectl get ingress "${KBS_INGRESS_NAME}" -n "${KBS_NS}" -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+			[[ -z "${host}" && ${SECONDS} -lt 30 ]] || break
+			sleep 5
```
Do you need to do SECONDS += 5 here? At the moment I think it will loop forever if there is an issue with ingress coming up?

wainersm replied:
Rá! This is a magic that I learned with @ldoktor while ago: that's a bash variable which is updated each time it's referenced. Quoting bash's manpage:

```
SECONDS
              Each time this parameter is referenced, it expands to the number
              of  seconds  since  shell invocation.  If a value is assigned to
              SECONDS, the value returned upon subsequent  references  is  the
              number  of seconds since the assignment plus the value assigned.
              The number of seconds at shell invocation and the  current  time
              are  always determined by querying the system clock.  If SECONDS
              is unset, it loses its special properties, even if it is  subse‐
              quently reset.
```

sprt reviewed (approved):
Thank you @wainersm!

sprt commented:
Force-merged as a workflow change tested in a feature.
